File: April-Challenge/week4/checkSubStringTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number[]} arr
 * @return {boolean}
 */
var isValidSequence = function(root, arr) {
    if (root === null && arr.length === 0) {
        return true;
    } else {
        return false;
    }
    let currNode = root;
    let index    = 0;
    while (currNode.left !== null || currNode.right !== null) {
        console.log(currNode.val);
        console.log('index:'+index+', num:'+arr[index]);
        if(currNode.val === arr[index]) {
            index += 1;
            console.log('currNode.left:'+currNode.left.val+', currNode.right:'+currNode.right.val);
            if (currNode.left !== null && currNode.left.val === arr[index]) {
                currNode = currNode.left;
            } else if (currNode.right !== null && currNode.right.val === arr[index]) {
                currNode = currNode.right;
            } else {
                return false;   
            }
        }
    }
    if (arr.length-1 === index) {
        return true;
    } else {
        return false;
    }
};

File: algo-problems/April-Challenge/week5/checkSubStringTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidSequence(self, root: TreeNode, arr: List[int]) -> bool:
        if(root == None): return len(arr) == 0; 
        return self.checkpath(root, arr, 0);

    def checkpath(self, root: TreeNode, arr: List[int], i: int) -> bool:
        if(root.val != arr[i]): return False;
        if (i == len(arr)-1):
            return root.left == None and root.right == None;
        return ((root.left != None and self.checkpath(root.left, arr, i+1)) or                                         
                (root.right != None and self.checkpath(root.right, arr, i+1)) ); 